Log out from multiple target portals when using iscsi storage plugin

mtanino opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

When using iscsi storage with multiple target portal (TP) addresses
and multipathing the volume manager logs on to the IQN for all
portal addresses, but when a pod gets destroyed the volume manager
only logs out for the primary TP and sessions for another TPs are
always remained.

This patch adds mount points for all TPs, and then log out from all
TPs when a pod is destroyed. If a TP is referred from another pods,
the connection will be remained as usual.



**Which issue this PR fixes** 
fixes #45394

**Special notes for your reviewer**:

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
iscsi storage plugin: Fix dangling session when using multiple target portal addresses.
```

gnufied commented:
/assign @rootfs

rootfs commented:
@k8s-bot ok to test

tobad357 commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -390,3 +407,31 @@ func extractPortalAndIqn(device string) (string, string, error) {
+			portals = append(portals, tp)
+		}
+	}
+	return portals
```
My go fu is not great but this looks like it will return an array with empty elements at the end in case it removes some duplicates which would mean the kubelet tries to execute iscsiadm commands on empty data.

This go playground example avoids it, maybe can use that
https://play.golang.org/p/q3bZ3hpOzD

mtanino replied:
Thank you for your comment. Your suggestion looks good.
My go is not great too, so I'm curious what kind of input causes empty elements return?

tobad357 replied:
My bad
I expected the len of the new array to be 3 and the last element empty in the case
of having a list of three portals with on set of duplicates

See
https://play.golang.org/p/hkm8_bhlOy

tobad357 commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -390,3 +407,31 @@ func extractPortalAndIqn(device string) (string, string, error) {
+	return portals
+}
+
+// If all mntPaths indicate global mounts, then return ture. Otherwise return false.
```
Nit: ture = true

tobad357 commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -390,3 +407,31 @@ func extractPortalAndIqn(device string) (string, string, error) {
+			return false
+		}
+	}
+	return flag
```
can just return true here as flag isn't used anywhere

mtanino replied:
Correct. Thanks.

tobad357 commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -263,66 +254,92 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
+						glog.Errorf("iscsi: failed to delete node record Error: %s", string(out))
+					}
+				} else {
+					glog.Infof("iscsi: log out target %s iqn %s", portal, iqn)
```
@rootfs how long backwards compatibility is promised.
Can this code be removed after X versions of k8s.
If so would be good to add a todo for when it should be removed

rootfs replied:
a todo is good for note taking

mtanino replied:
OK. Will do.

tobad357 commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -263,66 +254,92 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
+
+	for _, mntPath := range mntPaths {
+		// if device is no longer used, see if need to logout the target
+		if cnt == 0 {
```
I'm not quite following the logic here
If there is two mntpaths where the first one has a cnt of 0 and the first one the second one a count of 2.
Won't we skip logging out the first one?

mtanino replied:
Agreed. This logic should be changed like following.

In case of multipath with two portals, mntpaths has two global mount paths like;
     /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/x.x.x.10:3260-iqn.aaa
     /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/x.x.x.20:3260-iqn.bbb

THen, we should;
(1) Unmount all mntPaths
(2) Count num(cnt) of remained devices with GetDeviceNameFromMount
(3) If cnt == 0,then logout from these two iSCSI sessions

Any concerns?

tobad357 replied:
Is there a reason we need to wait to log out one device until the other one has cnt == 0
If we don't have to wait and just log out any device with cnt == 0 then we can combine the loops

mtanino replied:
The cnt shows that the number of remained devices
If we have two remaining paths like this, then GetDeviceNameFromMount returns '2'.

/var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/x.x.x.10:3260-iqn.aaa
/var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/x.x.x.20:3260-iqn.bbb

In this case, I think we should unmount these paths at first. Then if these unmount is succeeded and cnt == 0, it means no one is using iSCSI connection so we can logout each iSCSI connections step by step. so I'm using divided for two loop for these steps.

But can we combine these steps into single for loop? Please let me know if you have good idea.
Thanks,

tobad357 commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -263,66 +254,92 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
-			// This portal/iqn/iface is no longer referenced, log out.
-			// Extract the portal and iqn from device path.
-			portal, iqn, err := extractPortalAndIqn(device)
+		if err = c.mounter.Unmount(mntPath); err != nil {
```
By doing Unmount first and GetDeviceNameFromMount second can't the cnt, cnt-- be skipped
and only use cnt

mtanino replied:
Will do.

mtanino commented:
ok to test

mtanino commented:
@rootfs
Can you take a look this fix?

mtanino commented:
@k8s-bot pull-kubernetes-cross test this

rootfs commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -238,91 +246,132 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
+		if len(bkpPortal) == 1 {
+			globalPDPath = b.manager.MakeGlobalPDName(*b.iscsiDisk, "")
+		} else {
+			//
```
empty line

mtanino replied:
I'll remove it.

rootfs commented:
@humblec

humblec commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -167,7 +171,11 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
 	iscsiTransport = extractTransportname(string(out))
 
-	bkpPortal := b.portals
+	bkpPortal := removeDuplicate(b.portals)
```
Why we need `removeDuplicate` in attachdisk, can you please give some details on why we need this ?

mtanino replied:
At first I thought if user passes duplicate portal IPs like this, iscsiadm will be called multiple times unnecessarily, but it was my misunderstanding. 
```
    iscsi:
      targetPortal: 192.168.122.85:3260
      portals: ['192.168.122.85:3260', '192.168.122.7:3260']
      iqn: iqn.2017-05.com.example:rhel7
```
Actually waitForPathToExist() checks existing path so even if user specify duplicate portal IPs in the config, duplicate IPs will be skipped.

I'll remove removeDuplicate() logic from attachdisk()

Thanks,

humblec commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -167,7 +171,11 @@ func (util *ISCSIUtil) AttachDisk(b iscsiDiskMounter) error {
 
-	bkpPortal := b.portals
+	bkpPortal := removeDuplicate(b.portals)
+	if len(bkpPortal) == 0 {
```
Its perfectly valid that portal can be empty. I may be missing the logic here, can you please fill me in?

mtanino replied:
same above.

humblec commented on pkg/volume/iscsi/disk_manager.go:
```diff
@@ -27,16 +27,16 @@ import (
 // Abstract interface to disk operations.
 type diskManager interface {
-	MakeGlobalPDName(disk iscsiDisk) string
+	MakeGlobalPDName(disk iscsiDisk, portal string) string
```
`portal` is a known member here in this code path as []string. May be you could use some other name to avoid confusion. Just a thought.

humblec commented:
@mtanino Thanks for this PR ! I have few comments.  Also it would be appreciated if you can explain some more on the `bind mount` approach followed in attachdisk for better understanding, I failed to get it in quick look.

humblec commented:
@mtanino : The test failure looks ot be valid as well.

```
I0527 08:03:19.329] pkg/volume/util/device_util.go:31: cannot use deviceHandler literal (type *deviceHandler) as type DeviceUtil in return argument:
I0527 08:03:19.329] 	*deviceHandler does not implement DeviceUtil (missing FindDevicemapperName method)
I0527 08:03:19.329] !!! [0527 08:03:19] Call tree:
I0527 08:03:19.330] !!! [0527 08:03:19]  1: /go/src/k8s.io/kubernetes/hack/lib/golang.sh:690 kube::golang::build_binaries_for_platform(...)
I0527 08:03:19.330] !!! [0527 08:03:19]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
I0527 08:03:19.336] Makefile:89: recipe for target 'all' failed
I0527 08:03:19.339] Makefile:396: recipe for target 'cross' failed
```

humblec commented:
@mtanino also the PR says, logout from multiple target portals when using iscsi plugin, however it change the code in attachdisk()..etc, where my expectation was that, the change would be more or less in detachdisk(). It could be better if you can split this PR into different commits and organize the changes. It will make the review process bit easier. At time of merging we could squash it .

mtanino commented:
@humblec
>> where my expectation was that, the change would be more or less in detachdisk().

In my understanding, 
- Currently only first path is mounted under kubelet dir even if user specify multiple portals.
```
/dev/mapper/mpathb on /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/192.168.122.85:3260-iqn.2017-05.com.example:rhel7-lun-0 
```
- DetachDisk() collects iSCSi connection information from this mount path.
  But we don't have 2nd and subsequent sessions information so that we can't logout from them.

Is this correct?

In my approach, In order to logout from multiple iSCSI sessions, we need following two fix in this PR.
- Add mount points for all iSCSI sessions at Attachdisk()
- Logout from multiple portals using above mount points information at DetachDisk()

As a result, if user specify multiple portals like this,
```
    iscsi:
      targetPortal: 192.168.122.85:3260
      portals: ['192.168.122.7:3260']
      iqn: iqn.2017-05.com.example:rhel7
```
we can see mount points like this.

```
/dev/mapper/mpathb on /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/192.168.122.85:3260-iqn.2017-05.com.example:rhel7-lun-0 type ext4 (ro,relatime,block_validity,delalloc,barrier,user_xattr,acl,stripe=2048)
/dev/mapper/mpathb on /var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/192.168.122.7:3260-iqn.2017-05.com.example:rhel7-lun-0 type ext4 (ro,relatime,block_validity,delalloc,barrier,user_xattr,acl,stripe=2048)
/dev/mapper/mpathb on /var/lib/kubelet/pods/f3838bc8-4562-11e7-8fd5-525400645a80/volumes/kubernetes.io~iscsi/iscsipd-rw type ext4 (ro,relatime,block_validity,delalloc,barrier,user_xattr,acl,stripe=2048)
```
Then we can properly logout using these two iSCSI connection information.

Any thought?

Thanks,
Mitsuhiro

mtanino commented:
@humblec 
I'll take a look device_util.go as well.

humblec commented:
@mtanino thanks for giving more details about the approach! iirc,  when there are entries in `portals` field, we form paths for each portal. You could validate this in `/dev/disk/by-path`. Please pay attention to the portal IPs ( 10.0.2.{15,16,17,18} and also the IQN. These ( as in below output)  all belongs to same `mpath` device. Also, we could list the active sessions from `iscsiadm -m session` .  I think once we confirmed there is no active mount/ref exist on this mpath device,  we can make use of existing iscsi sessions or parse below path to understand the existing connections and logout accordingly.  
 
```
# ll /dev/disk/by-path/
lrwxrwxrwx. 1 root root 9 Feb 16 15:58 ip-10.0.2.15:3260-iscsi-iqn.2016-04.test.com:storage.target00-lun-0 -> ../../sdb
lrwxrwxrwx. 1 root root 9 Feb 16 15:41 ip-10.0.2.16:3260-iscsi-iqn.2016-04.test.com:storage.target00-lun-0 -> ../../sdc
lrwxrwxrwx. 1 root root 9 Feb 16 15:58 ip-10.0.2.17:3260-iscsi-iqn.2016-04.test.com:storage.target00-lun-0 -> ../../sdd
lrwxrwxrwx. 1 root root 9 Feb 16 15:41 ip-10.0.2.18:3260-iscsi-iqn.2016-04.test.com:storage.target00-lun-0 -> ../../sde
```

@rootfs Any thoughts?

mtanino commented:
@humblec 

Yeah.
Originally I thought that steps but seems the logic is little bit confusing, this way needs analyzing of /proc/mounts, /sys/block/dm-x, /dev/disk/by-path, etc...

If we use /dev/disk/by-path information,
- at first, we only know mount path like this, 
```
/var/lib/kubelet/plugins/kubernetes.io/iscsi/iface-default/192.168.122.85:3260-iqn.2017-05.com.example:rhel7-lun-0
```
therefore, we need to get device file name using mount.GetDeviceNameFromMount(). we could get mpathX from the method.
- then we need to map this to dm-x reffering /sys/block/dm-X/dm/name(this contains mpath name).
- After that we need to find slave SCSI devices from /sys/block/dm-X/slaves/*.
```
ex.
% pwd
/sys/block/dm-7
% ls slaves
sda  sdb
```
- now we can get iscsi connection information from /dev/disk/by-path/

```
% ls -la /dev/disk/by-path
total 0
drwxr-xr-x 2 root root 180 May 30 15:43 .
drwxr-xr-x 6 root root 120 May 26 15:43 ..
lrwxrwxrwx 1 root root   9 May 30 15:43 ip-192.168.122.7:3260-iscsi-iqn.2017-05.com.example:rhel7-lun-0 -> ../../sdb
lrwxrwxrwx 1 root root   9 May 30 15:43 ip-192.168.122.85:3260-iscsi-iqn.2017-05.com.example:rhel7-lun-0 -> ../../sda
```
My quick question is, are there a way to get "iface" from these information?

humblec commented:
@mtanino One quick comment, if we get the `mpath` device name and once we are sure this mpath is no longer mounted  in the kubelet, we could just logout from sessions using volume spec's portal IPs and IQN, Isnt it ? we dont need to  parse `/sys/block/dm-x, /dev/disk/by-path, etc...`, ? 

Regarding# `My quick question is, are there a way to get "iface" from these information?`
I think its not possible, but I could be wrong.

mtanino commented:
@humblec 
Can we get portal IPs and IQN in the volume spec's when detaching volume?
I tried to debug them at DetachDisk()'s path but they were empty.

c iscsiDiskUnmounter
- c.portals
- c.iqn
- c.iface

I thought they were not stored into the etcd, therefore I made a logic to retrieve iSCSI connection information from mount path. But if we cloud get them from volume spec's, we can make the logic simpler.

tobad357 commented:
Another way to fix this would be to log out of all portals at once.
Looking at it I can't really see a reason why the kubelet would want to hang on to some sessions and not others if the cnt == 0 across all mounts

To log out all portals the command is the same but exclude -p <portal ip>
Something like this should work
iscsiadm -m node -T <<iqn>> -u

humblec commented:
@mtanino True. if we want to make use of the spec, we need to store and fetch in detachdisk(). Also, if we know the mpath name, we can fetch the paths from `multipath -ll` command ..etc.  As a second thought, if we are sure there is no mount exist, we could blindly logout from all sessions as @tobad357 said. But consider that, there is a possibility of one other pod mounted the same share/IQN in `RO` mode in the same system.

tobad357 commented:
@rootfs I wasn't aware that one could mount the same IQN via different portals as RW and RO.
Even if rare, could still cause some weird issues if blindly logging out all the portals.

@mtanino 
I think you could do the following which avoids going through all the DM devices.

1. mount.GetDeviceNameFromMount() //Get the device name like mpathX
2. Read the DM from /dev/mapper/mpathX
3. Read all the slave devices from /sys/block/dm-X/slaves. I have some old code that does that, see below
4. Filter out all the iqn and portals from /dev/disk-by/path

```
//FindDevicesForMultipathDevice given a disk name of /dev/dm-XX return a
//slice including all the mapped devices
//if the device isn't a mapped device return an empty slice
func (handler *osIOHandler) FindDevicesForMultipathDevice(device string) []string {
	disk, err := handler.FindDeviceForPath(device)
	if err != nil {
		return make([]string, 0, 0)
	}
	sysPath := "/sys/block/" + disk + "/slaves/"
	if dirs, err := handler.ReadDir(sysPath); err == nil {
		devs := make([]string, len(dirs), len(dirs))
		for i, f := range dirs {
			devs[i] = f.Name()
		}
		return devs
	}
	return make([]string, 0, 0)
}
```

mtanino commented:
@tobad357 @humblec 

>>To log out all portals the command is the same but exclude -p 
>>Something like this should work
>>iscsiadm -m node -T <> -u

I think this causes a problem when we have same IQN on multiple different backend storages like this.
Without specifying "-p" option, kubelet logouts from following two portals at once even if user only deletes a volume with pod on Srotage A. 

```
Storage A -> Portal IP: 192.168.122.100, iqn: iqn.2017-05.com.example:fc25 =>  sda, sdb, mpatha
Storage B -> Portal IP: 10.20.30.200,    iqn: iqn.2017-05.com.example:fc25  => sdc
```

>> Even if rare, could still cause some weird issues if blindly logging out all the portals.

Yes. might be rare case but could cause above issue.

Also thank you for pointing out the steps to retrieve DM and by-path.
As for the another choice, I got a suggestion which stores iscsiDiskMounter information into json file under globalPDPath dir from @rootfs same as RBD.

https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/rbd/rbd_util.go#L195

I'm considering this way might be simpler and robust compared to analyzing command results and /sys and /dev. But I'm still under investigation.

tobad357 commented:
@mtanino if going down the route of saving a json file wont that become an issue in the scenario where the umount goes through but something happens with the iscsi logout.
In that case the kubelet will not be able to logout the sessions in successive tries

rootfs commented:
@tobad357 the json file lives on the host, so it will survive an unmount and iscsi logout.

Commit:
Log out from multiple portals with iscsi storage plugin

When using iscsi storage with multiple target portal (TP)
addresses and multipathing the volume manager logs on to
the IQN for all portal addresses, but when a pod gets
destroyed the volume manager only logs out for the primary
TP and sessions for another TPs are always remained.

This patch adds methods to store and load iscsi disk
configrations, then uses the stored config at DetachDisk
path.

Fix #45394

rootfs commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -150,7 +151,35 @@ func makePDNameInternal(host volume.VolumeHost, portal string, iqn string, lun s
+	file := path.Join(mnt, "iscsi.json")
+	fp, err := os.Create(file)
+	if err != nil {
+		return fmt.Errorf("iscsi: create err %s/%s", file, err)
```
`iscsi: create err %s/%s`

mtanino replied:
done

rootfs commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -150,7 +151,35 @@ func makePDNameInternal(host volume.VolumeHost, portal string, iqn string, lun s
+	file := path.Join(mnt, "iscsi.json")
+	fp, err := os.Open(file)
+	if err != nil {
+		return fmt.Errorf("iscsi: open err %s/%s", file, err)
```
ditto

mtanino replied:
done

rootfs commented on pkg/volume/iscsi/iscsi_util.go:
```diff
@@ -150,7 +151,35 @@ func makePDNameInternal(host volume.VolumeHost, portal string, iqn string, lun s
+	return nil
+}
+
+func (util *ISCSIUtil) loadISCSI(conf *iscsiDisk, mnt string) error {
```
nit: make a note here, the iscsi config json is not deleted

mtanino replied:
done

rootfs commented:
/lgtm

mtanino commented:
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

mtanino commented:
@k8s-bot pull-kubernetes-node-e2e test this

humblec commented:
/lgtm. Thanks !!

mtanino commented:
After merged the fix, can we cherry pick the PR to v1.6.x tree?

childsb commented:
/approve

mtanino commented:
/cherrypick-candidate

mtanino commented:
cherrypick-candidate

tobad357 commented:
@mtanino I agree this is very important to us as well
Would be great to get into a 1.6.x release

mtanino commented:
@tobad357 

Yes, but I'm not clearly understand the sequence of cherry-pick mentioned below, especially the way to add cherrypick-candidate label.
https://github.com/kubernetes/community/blob/master/contributors/devel/cherry-picks.md
